Add documentation for review queues and label schemas (#23975)
mlflow/mlflow: docs/docs/genai/assessments/review-queues.mdx

---
sidebar_label: Review Queues
description: Organize traces into review queues, assign them to human reviewers, and collect structured feedback and ground-truth annotations in the MLflow UI.
---

import FeatureHighlights from "@site/src/components/FeatureHighlights";
import ConceptOverview from "@site/src/components/ConceptOverview";
import { APILink } from "@site/src/components/APILink";
import { Users, ListCheck, MessageSquare, Database, Target, TrendingUp } from "lucide-react";

# Review Queues

Review queues turn ad-hoc human review into a structured workflow. You bundle a set of traces into a named queue, attach the questions you want answered, assign the queue to one or more reviewers, and let them work through it in focus mode. Every answer is logged back onto the trace, so the review effort feeds directly into evaluation.

:::note Experimental
Review queues are experimental (added in MLflow 3.14.0) and may change in a future release.
:::

> ## What is a Review Queue?

Reviewer: Can we consolidate this with "Why use Review Queues?". They look alike and reads a bit repetative.

Author: Consolidated. Merged the two into a single "What is a Review Queue?" section, folding the value props into the intro so the card lists no longer repeat each other.

-- Claude-drafted, reviewed before posting.

A review queue is a named collection of traces, scoped to an experiment, that a defined set of reviewers works through together. Three building blocks make up the workflow:

<ConceptOverview concepts={[
  {
    icon: ListCheck,
    title: "Review Questions",

>     description: "The questions reviewers answer on each trace — a Pass/Fail toggle, a categorical choice, a numeric score, or free text."

Reviewer: Let's not use em-dash. Many people see this as a sign of AI slop.

Author: Done. Scrubbed every em dash from the page (replaced with commas, colons, or parentheses).

-- Claude-drafted, reviewed before posting.

  },
  {
    icon: Users,
    title: "Review Queues",
    description: "A bundle of traces, the questions to ask, and the reviewers assigned to work through them."
  },
  {
    icon: MessageSquare,
    title: "Shared Review Status",
    description: "Each trace is pending, complete, or declined — shared across reviewers, so the first to finish a trace clears it for the whole team."
  }
]} />

## Prerequisites

- MLflow 3.14.0 or later.
- An MLflow tracking server (or local tracking setup) with traces already logged to an experiment.

> - For multi-reviewer (custom) queues and per-reviewer attribution, run the tracking server with [authentication](/self-hosting/security/basic-http-auth) enabled. On a server without authentication, reviews are attributed to a single `default` user and reviewer assignment is hidden.

Reviewer: nit: `"For multi-reviewer (custom) queues and per-reviewer attribution"`: it's not clear what this means. Can we say "To assign reviews using user name"?

Reviewer: Also let's link to the section (#create-a-custom-queue)

Author: Reworded to "To assign reviews using a user name" and linked to the [custom queue](#create-a-custom-queue) section from the prerequisite.

-- Claude-drafted, reviewed before posting.

## Why use Review Queues?

<FeatureHighlights features={[
  {
    icon: Target,
    title: "Structured, Consistent Review",
    description: "Reviewers answer the same well-defined questions on every trace instead of leaving free-form notes, so results are comparable and aggregatable."
  },
  {
    icon: Users,
    title: "Multi-reviewers Collaboration",
    description: "Assign a queue to several reviewers and let the shared review status keep everyone on different traces, with clear attribution of who reviewed what."
  },
  {
    icon: Database,
    title: "Feeds Evaluation Directly",
    description: "Every answer is logged back onto the trace, building ground-truth datasets and quality signals you can reuse in evaluation."
  },
  {
    icon: TrendingUp,
    title: "Track Progress at a Glance",
    description: "A progress bar and per-status counts show how much of a queue is done, so you know when a labeling effort is complete."
  }
]} />

## Assign traces to reviewers

> The quickest way to get traces reviewed: from the trace table (or an open trace), select the traces, click **Assign to reviewers**, and pick one or more reviewers. Each trace lands in that reviewer's personal queue, where they answer the experiment's review questions. (On a server without authentication, reviewer assignment isn't available — use a custom queue instead, described below.)

Reviewer: ```suggestion
The quickest way to get traces reviewed: from the trace table (or an open trace), select the traces, click **Assign to reviewers**, and pick one or more reviewers. Each trace lands in that reviewer's personal queue, where they answer the experiment's review questions. (On a server without authentication, direct reviewer assignment isn't available. Use a custom queue instead, described below.)
```

Author: Applied your suggestion verbatim.

-- Claude-drafted, reviewed before posting.

![Assigning traces to reviewers in the MLflow UI](/images/assessments/review-queue-assign-to-reviewers.png)

## Review traces in the UI

Reviewers find the work assigned to them under the experiment's **Review** tab, where each queue shows a progress bar across its traces.

> ![The Review tab listing queues and progress](/images/assessments/review-queue-tab-overview.png)

Reviewer: Can we replace this screenshot with non-empty queue?

Author: Done. Recaptured with the Ground Truth queue selected so the trace list and status counts show, instead of the empty "Select a queue" state.

-- Claude-drafted, reviewed before posting.

> Selecting a trace opens **focus mode** — a full-page surface with the queue's traces on the left, the trace's input and output in the middle (with **View full trace** for the complete trace explorer), and the review questions on the right.

Reviewer: ```suggestion
Selecting a trace opens the focused review page — a full-page surface with the queue's traces on the left, the trace's input and output in the middle (with **View full trace** for the complete trace explorer), and the review questions on the right.
```

nit: "focus mode" is our internal name, we don't name it in the product

Author: Took the rename and dropped that internal name from the docs. One deviation: the page no longer has a left trace-list panel, so I reworded to "input and output in the main pane and the review questions on the right" to match the current two-pane layout instead of describing traces on the left. Let me know if I have misread the current UI.

-- Claude-drafted, reviewed before posting.

![Reviewing a trace in focus mode](/images/assessments/review-queue-focus-mode.png)

In focus mode a reviewer can:

- **Submit** their answers to mark the trace complete (a single Pass/Fail question submits as soon as it's picked).
- **Decline** a trace that was added to the queue by mistake.
- Re-open a completed answer, edit it, and **Save changes** in place — or use **Move to Todo** to send the trace back to the pending pool.

Because the status is shared across the queue, the first reviewer to complete a trace clears it for everyone, with `completed_by` recording who did it.

## Create a custom queue

When you want a curated, shared queue that asks a **specific set of questions** — or you're on a server without authentication — create a custom queue instead of assigning to individual reviewers.

> **1. Define your review questions.** From the experiment's **Review** tab, open the create-queue dialog and click **New question**:

Reviewer: ```suggestion
#### Step 1. Define your review questions.

From the experiment's **Review** tab, open the create-queue dialog and click **New question**:
```

nit: for better layout
<img width="775" height="231" alt="Screenshot 2026-06-14 at 22 35 06" src="https://github.com/user-attachments/assets/3243d46b-47f1-495e-b2a0-6b7b228e252b" />

Author: Applied; the questions step is now a `#### Step 1.` heading.

-- Claude-drafted, reviewed before posting.

1. Name the question.
2. Pick an input type — Pass/Fail, single- or multi-select categorical, numeric, or free text.
3. Optionally add an instruction shown to reviewers, and allow a free-text comment.
4. Choose whether it's a **feedback** question (how good was the output?) or an **expectation** question (what should the output have been?).

Questions are defined once and reused across the queues in that experiment.

> ![Creating a review question in the MLflow UI](/images/assessments/review-queue-new-question.png)

Reviewer: Let's use `ImageBox`. Without border, these screenshots are hard to distinguish from the main body.

Author: Done. All five screenshots are now wrapped in `<ImageBox>` for the border.

-- Claude-drafted, reviewed before posting.

> **2. Create the queue.** Name it, pick the questions it should ask, assign reviewers (on an authenticated server), and click **Create**. Then add traces to it with **Assign to reviewers** from the trace table.

Reviewer: nit: Can we separate "Assign to reviewers" as third step? Although it's in the create modal, they are decoupled action indeed.

1. Create questions.
2. Create a queue.
3. Assign the queue to reviewers.

Author: Split into three steps: 1) define questions, 2) create the queue, 3) assign reviewers and add traces.

-- Claude-drafted, reviewed before posting.

:::note
`InputCategoricalList`, `InputTextList`, and some input options are only fully supported on Databricks-managed tracking. On Databricks, `create_label_schema` routes to the Databricks Review App; on open-source MLflow it uses the tracking store.
:::

## How review answers get logged to traces

A review queue never stores answers separately from your traces. Each answer is written straight onto the reviewed trace — a **feedback** question logs feedback, an **expectation** question logs the expected (ground-truth) value — attributed to the reviewer. Re-saving an edited answer replaces the reviewer's previous one rather than creating a duplicate. This means review output is immediately available wherever trace data is used — the trace UI, [evaluation](/genai/eval-monitor), and ground-truth [datasets](/genai/datasets) — with no extra export step.

> ## API Reference

Reviewer: nit: I think we don't need this list. These will be niche use case and users can go to API doc from the "Manage and track queues programmatically" section.

Author: Removed the API Reference list. The "Manage and track queues programmatically" section already links into the API docs for the niche calls.

-- Claude-drafted, reviewed before posting.

> 

Reviewer: Instead, can we add "Next step" section just like other pages? As we discussed, it is important users have a clear path for using logged feedback, such as evaluation, align judges, creating a dataset (ground truth).

Author: Added a "Next steps" section linking evaluation, judge alignment, and building an evaluation dataset.

-- Claude-drafted, reviewed before posting.

---
sidebar_label: Review Queues
description: Organize traces into review queues, assign them to human reviewers, and collect structured feedback and ground-truth annotations in the MLflow UI.
---

import FeatureHighlights from "@site/src/components/FeatureHighlights";
import ConceptOverview from "@site/src/components/ConceptOverview";
import CollapsibleSection from "@site/src/components/CollapsibleSection";
import { APILink } from "@site/src/components/APILink";
import { Users, ListCheck, MessageSquare, Database, Target, TrendingUp, Eye } from "lucide-react";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import TabsWrapper from "@site/src/components/TabsWrapper";

> # Review Queues

Reviewer: Can we re-factor the existing "Feedback Collection" page significantly to focus on the new review flow, instead of adding a new review queue page?

https://pr-23975--mlflow-docs-preview.netlify.app/docs/latest/genai/assessments/feedback/

The point of that page is "how to get human review in MLflow", and it is exactly where we want to evangelize what we've built. Honestly, the current "Feedback Collection" page includes lots of implementation details and does not show a good overview on the user flow, so feel free to drop them or outsource them to API doc😅

Reviewer: Tho if you think this needs more consideration, feel free to merge this as-is and happy to discuss offline.

One potential structure is

```
- Review and Annotation
   - Collaborative Human Review (the 'assign' flow)
   - Collecting End User Feedback (programatic log_feedback)
   - Annotate Ground Truth for Evaluation (expectation -> dataset flow)
```

Author: For this PR I kept it as a standalone page so the content ships, and I'd suggest the broader "Review and Annotation" reorg (folding in the Feedback Collection refactor) as a separate follow-up, since it touches the existing feedback/expectations pages. Happy to take that offline as you offered.

-- Claude-drafted, reviewed before posting.

Review queues turn ad-hoc human review into a structured workflow. You bundle a set of traces into a named queue, attach the questions you want answered, assign the queue to one or more reviewers, and let them work through it in focus mode. Every answer is recorded back on the trace as a [feedback](/genai/assessments/feedback) or [expectation](/genai/assessments/expectations) assessment, so the review effort feeds directly into evaluation and monitoring.

For the full API, see the <APILink fn="mlflow.genai.review_queues.create_review_queue">`mlflow.genai.review_queues`</APILink> and <APILink fn="mlflow.genai.label_schemas.create_label_schema">`mlflow.genai.label_schemas`</APILink> references.

> :::note Experimental

Reviewer: <img width="641" height="78" alt="Screenshot 2026-06-14 at 19 14 32" src="https://github.com/user-attachments/assets/a873afd5-fbf8-46f8-9928-0802678d50f5" />

nit: rendering is broken

Author: Good catch — I restructured the top of the page and removed the line that was rendering oddly (the intro no longer has an API-link line, just the experimental admonition). Could you re-check on the latest preview once it rebuilds?

-- Claude-drafted, reviewed before posting.

Reviewer: @kriscon-db this is not fixed.

Author: Fixed for real this time. The admonition was missing the bracket syntax; it is now `:::note[Experimental]`, which renders as a titled note. Please re-check on the rebuilt preview.

-- Claude-drafted, reviewed before posting.

The review-queue and review-question APIs are experimental (added in MLflow 3.14.0) and may change in a future release.
:::

> ## What is a Review Queue?

Reviewer: This section is a bit too long text to start with. Can we simplify to  bullet points of one short sentence per each, or maybe replace it with a tabs of screenshot + a short sentence?

Author: Done — trimmed each concept card to a single short sentence.

-- Claude-drafted, reviewed before posting.

  },
  {
    icon: MessageSquare,
    title: "Shared Review Status",
    description: "Each trace in a queue is pending, complete, or declined, and that status is shared across the queue's reviewers — the first reviewer to finish a trace marks it complete for the whole queue — so a team can divide and conquer without double-work."
  }
]} />

## Prerequisites

- MLflow 3.14.0 or later.
- An MLflow tracking server (or local tracking setup) with traces already logged to an experiment.
- For multi-reviewer (custom) queues and per-reviewer attribution, run the tracking server with [authentication](/self-hosting/security/basic-http-auth) enabled. On a server without authentication, reviews are attributed to a single `default` user and reviewer assignment is hidden.

## Why use Review Queues?

<FeatureHighlights features={[
  {
    icon: Target,
    title: "Structured, Consistent Review",
    description: "Reviewers answer the same well-defined questions on every trace instead of leaving free-form notes, so results are comparable and aggregatable."
  },
  {
    icon: Users,

>     title: "Divide Work Across a Team",

Reviewer: ```suggestion
    title: "Multi-reviewers Collaboration",
```

nit: I've seen feature request asks "multi-reviewer" explicitly, so we can surface it more directly.

Author: Applied — retitled to "Multi-reviewers Collaboration".

-- Claude-drafted, reviewed before posting.

    description: "Assign a queue to several reviewers and let the shared review status keep everyone on different traces, with clear attribution of who reviewed what."
  },
  {
    icon: Database,
    title: "Feeds Evaluation Directly",
    description: "Every answer is written back to the trace as a feedback or expectation assessment, building ground-truth datasets and quality signals you can reuse in evaluation."
  },
  {
    icon: TrendingUp,
    title: "Track Progress at a Glance",
    description: "A progress bar and per-status counts show how much of a queue is done, so you know when a labeling effort is complete."
  }
]} />

> ## Step 1: Define your review questions

Reviewer: The default flow we want to push is "One click to flag traces to reviewers" (auth-enabled path). Can we use it as a primary simplest path, and put the queue and question creation as a custom queue user flow (or workaround for users who don't enable auth)?

Author: Done — reordered so the one-click "Assign to reviewers" flow is the primary path, then "Review traces in the UI", with queue + question creation demoted to a "Create a custom queue" section (also the no-auth path).

-- Claude-drafted, reviewed before posting.

      - A custom queue's questions (`schema_ids`) are frozen once it has traces attached. Detach the traces first if you need to change the questions; the assigned `users` stay editable at any time.
    </div>

  </TabItem>
</Tabs>
</TabsWrapper>

## Step 3: Add traces to a queue

Attach the traces you want reviewed to the queue.

<TabsWrapper>
<Tabs>
  <TabItem value="ui" label="UI" default>
    <div class="flex-column">
      1. From the trace table (or an open trace), select the traces you want reviewed.
      2. Use **Assign to reviewers** to flag them into one or more queues.

      ![Assigning traces to a review queue in the MLflow UI](/images/assessments/review-queue-assign-to-reviewers.png)
    </div>

  </TabItem>
  <TabItem value="python" label="Python">
    <div class="flex-column">
      Attach traces by their trace IDs. Adding is idempotent — re-adding a trace keeps its existing review status.

      ```python
      from mlflow.genai.review_queues import add_items_to_review_queue

      add_items_to_review_queue(queue.queue_id, item_ids=["tr-abc123", "tr-def456"])
      ```
    </div>

  </TabItem>
</Tabs>
</TabsWrapper>

> ## Step 4: Review traces in the UI

Reviewer: Since we worked very hard to build UI that speaks itself, let's avoid describing all details with words, which might make the feature looks more complex to users than it is.

Author: Agreed — the walkthrough is now UI/screenshot-first (Python pulled out into a single SDK section at the bottom), so the screenshots carry the explanation instead of long prose.

-- Claude-drafted, reviewed before posting.

Reviewers work a queue from the experiment's **Review** tab. Selecting a trace opens **focus mode**, a full-page review surface with three panels:

- **Left** — a compact rail of the queue's traces, with a progress bar across the top, for moving between items.
- **Middle** — the trace's input and output. Chat-style and plain-text content render as formatted Markdown, JSON renders as a syntax-highlighted block, and **View full trace** opens the full trace explorer in a drawer.
- **Right** — the queue's review questions as input widgets (from Step 1), each with a status dot showing whether it has been answered.

![Reviewing a trace in focus mode](/images/assessments/review-queue-focus-mode.png)

> <CollapsibleSection title="The review workflow in focus mode" defaultExpanded={false}>

Reviewer: nit: Do we want-to/need collapse here? Review flow is important.

Author: Done — un-collapsed it, so the focus-mode review flow is now inline.

-- Claude-drafted, reviewed before posting.

    list_review_queue_items,
    set_review_queue_item_status,
    list_review_queues,
)

# How much of the queue is left?
pending = list_review_queue_items(queue.queue_id, status="pending")
print(f"{len(pending)} traces still to review")

# Mark a trace complete (completed_by is required for terminal states).
set_review_queue_item_status(
    queue.queue_id,
    item_id="tr-abc123",
    status="complete",
    completed_by="alice@company.com",
)

# Reopen it (clears completed_by).
set_review_queue_item_status(queue.queue_id, item_id="tr-abc123", status="pending")

# Find which queues a reviewer is assigned to.
my_queues = list_review_queues(user="alice@company.com")
```

You can also <APILink fn="mlflow.genai.review_queues.update_review_queue">`update_review_queue`</APILink> to change a custom queue's reviewers or questions, <APILink fn="mlflow.genai.review_queues.remove_items_from_review_queue">`remove_items_from_review_queue`</APILink> to detach traces, and <APILink fn="mlflow.genai.review_queues.delete_review_queue">`delete_review_queue`</APILink> to delete a queue. Deleting a queue leaves the reviewer assessments already recorded on its traces untouched.

> ## How review answers become assessments

Reviewer: ```suggestion
## How review answers get logged to traces
```

nit: I might be a bit obsessed to this, but I don't feel the concept "assessment" is widely adopted (at least I only see it in MLflow). So I suspect it's not straight-forward to understand what "feedback becomes assessment" means...

Author: Applied — renamed the heading to "How review answers get logged to traces".

-- Claude-drafted, reviewed before posting.
